io-engine/src/bdev/ftl.rs

```diff
@@ -0,0 +1,258 @@
+                uri: uri.to_string(),
+                message: String::from("No bbdev parameter found"),
+            })?
+            .to_string()
+            .replace("!", "?")
+            .replace("|", "&");
+
+        let cbdev_uri = parameters
+            .remove("cbdev")
+            .context(bdev_api::InvalidUri {
+                uri: uri.to_string(),
+                message: String::from("No cbdev parameter found"),
+            })?
+            .to_string()
+            .replace("!", "?")
+            .replace("|", "&");
+
+        // Device parameter checking for bbdev and cbdev are done in bdev_ftl_create_bdev
+        reject_unknown_parameters(uri, parameters)?;
+
+        Ok(Self {
+            name: uri.path()[1..].into(),
+            alias: uri.to_string(),
+            uuid,
+            bbdev_uri,
+            cbdev_uri,
+        })
+    }
+}
+
+impl GetName for Ftl {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+pub extern "C" fn ftl_bdev_init_fn_cb(
+    _ptr: *const ftl_bdev_info,
+    sender_ptr: *mut c_void,
+    errno: i32,
+) {
+    info!("{:?}: ftl_bdev_init_fn_cb", errno);
+    let sender = unsafe {
+        Box::from_raw(sender_ptr as *mut oneshot::Sender<ErrnoResult<()>>)
+    };
+    sender
+        .send(errno_result_from_i32((), errno))
+        .expect("done callback receiver side disappeared");
+}
+
+#[async_trait(?Send)]
+impl CreateDestroy for Ftl {
+    type Error = BdevError;
+
+    async fn create(&self) -> Result<String, Self::Error> {
+        if UntypedBdev::lookup_by_name(&self.name).is_some() {
+            return Err(BdevError::BdevExists {
+                name: self.name.clone(),
+            });
+        }
+
+        debug!(
+            "{:?}: Creating ftl-bdev '{:?}' with bbdev {:?} and cbdev {:?}.",
+            self,
+            CString::new(self.name.clone()).unwrap(),
+            self.bbdev_uri,
+            self.cbdev_uri
+        );
+
+        use std::ffi::CString;
+        extern crate libc;
+
+        let base_dev_name = bdev_create(&self.bbdev_uri).await?;
+        let cache_dev_name = bdev_create(&self.cbdev_uri).await?;
+
+        let l_name_ptr =
+            CString::new(self.name.clone()).unwrap().into_raw() as *mut i8;
+        let l_bname_ptr = base_dev_name.as_ptr() as *mut i8;
+        let l_cname_ptr = cache_dev_name.as_ptr() as *mut i8;
+
+        let (s, r) = oneshot::channel::<ErrnoResult<()>>();
+
+        let spdk_ftl_conf_size = mem::size_of::<spdk_ftl_conf>() as u64;
+        let mut ftl_conf = spdk_ftl_conf {
+            ..unsafe { mem::zeroed() }
+        };
+        unsafe { spdk_ftl_get_default_conf(&mut ftl_conf, spdk_ftl_conf_size) };
+        ftl_conf.name = l_name_ptr;
+        ftl_conf.base_bdev = l_bname_ptr;
+        ftl_conf.cache_bdev = l_cname_ptr;
+        ftl_conf.fast_shutdown = true;
+        ftl_conf.verbose_mode = true;
+        // SPDK_FTL_MODE_CREATE = (1 << 0)
+        ftl_conf.mode = 1 << 0;
```

Comment:
We don't have access to `SPDK_FTL_MODE_CREATE` ?

Reply:
Thanks for addressing this!
I tried importing `SPDK_FTL_MODE_CREATE` via `enum spdk_ftl_mode` which is located in `spdk/include/spdk/ftl.h` alongside `struct spdk_ftl_conf` which I am also using here.

However, when doing

```
use spdk_rs::{
    ffihelper::errno_result_from_i32,
    libspdk::{
        bdev_ftl_create_bdev, bdev_ftl_delete_bdev, ftl_bdev_info,
        spdk_ftl_conf, spdk_ftl_get_default_conf, spdk_ftl_mode,
    },
    UntypedBdev,
};
```

I am getting the following error:

```
error[E0432]: unresolved import `spdk_rs::libspdk::spdk_ftl_mode`
  --> io-engine/src/bdev/ftl.rs:44:51
   |
44 |         spdk_ftl_conf, spdk_ftl_get_default_conf, spdk_ftl_mode,
   |                                                   ^^^^^^^^^^^^^
   |                                                   |
   |                                                   no `spdk_ftl_mode` in `libspdk`
   |                                                   help: a similar name exists in the module: `spdk_ftl_conf`
```

What am I doing wrong here? I can't figure out what to change in spdk-rs' build.rs to include this enum.

Reply:
Do you need to add it to `allowlist_type` on the spdk-rs build.rs?

Reply:
Great, approved that one too!